### PR TITLE
AWS optimisations for getInstancesByNodeNames() and getCandidateZonesForDynamicVolume()

### PR DESCRIPTION
**Pull request by EricMountain**
* Lookup by instance-id in aws.go getInstancesByNodeNames()
* Use node-informer for getCandidateZonesForDynamicVolume() (eliminates describe-instances call to retrieve all running instances)
* Make DescribeInstances() log the AWS requestID on error

### CODE REVIEW

**Comment by EricMountain** [comment]
@lbernail on the optimised path I’m doing this: https://github.com/DataDog/kubernetes/pull/17/files#diff-07ba008af9c76b0539556ff7fde3105eR4289-R4299 - is this good enough? Or do you think that I need to query only the instance IDs first, and then filter out the instances that don’t match the states from the results I get back from AWS?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -2373,7 +2373,7 @@ func (c *Cloud) GetVolumeLabels(volumeName KubernetesVolumeID) (map[string]strin
 	labels := make(map[string]string)
 	az := aws.StringValue(info.AvailabilityZone)
 	if az == "" {
-		return nil, fmt.Errorf("volume did not have AZ information: %q", info.VolumeId)
+		return nil, fmt.Errorf("volume did not have AZ information: %s", *info.VolumeId)
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go]: Is this related to the patch?

**Comment by lbernail** [comment]
Looks ready for testing
Let's verify that it works as expected and create a PR upstream for discussion
What do you think?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -4238,9 +4238,76 @@ func (c *Cloud) getInstancesByIDs(instanceIDs []*string) (map[string]*ec2.Instan
...
+		glog.V(3).Infof("Failed to find any instances %v", nodeNames)
+		return nil, nil
+	}
+	return ec2Instances, nil
+}
+
+func (c *Cloud) getInstancesByInstanceIDs(instanceIDs []*string, states ...string) ([]*ec2.Instance, error) {
 	ec2Instances := []*ec2.Instance{}
+	for i := 0; i < len(instanceIDs); i += filterNodeLimit {
+		end := i + filterNodeLimit
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go]: I thik we need another constant (than filterNodeLimit) when retrieving by id because we can 1000+ instances with this method

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go]: Build of the tests fails due to more modern checks.

Comment (lbernail) [reply pkg/cloudprovider/providers/aws/aws.go]: Ok makes sense

File: pkg/cloudprovider/providers/aws/aws.go (hunk 2 above)
Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go]: Agreed, I need to look into how big we can make that.

Comment (lbernail) [reply pkg/cloudprovider/providers/aws/aws.go]: There is no official documentation but from informal discussion with AWS there does not seem to be a limit and 1000 is sure to work. Maybe use 1000 for a start?

Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go]: Done!

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -775,6 +775,9 @@ func (s *awsSdkEC2) DescribeInstances(request *ec2.DescribeInstancesInput) ([]*e
 		response, err := s.ec2.DescribeInstances(request)
 		if err != nil {
 			recordAwsMetric("describe_instance", 0, err)
+			if reqerr, ok := err.(awserr.RequestFailure); ok {
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:778]: Coud we also add the requestID in the success cases for GetById (in the current debug logs)?

Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:781]: As discussed this requires changes in the AWS SDK: the RequestID isn't available in response objects handed back through the API.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 4]
```diff
@@ -198,6 +199,8 @@ const ServiceAnnotationLoadBalancerHCTimeout = "service.beta.kubernetes.io/aws-l
 // service to specify, in seconds, the interval between health checks.
 const ServiceAnnotationLoadBalancerHCInterval = "service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval"
 
+const labelNodeRoleMaster = "node-role.kubernetes.io/master"
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:202]: I think this const is already available somewhere in the codebase (I've seen it before)

Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:202]: It's in a few places, but nothing we'd want to import in here AFAICT:

* `cmd/kubeadm/app/constants/constants.go`
* `pkg/controller/service/service_controller.go`

Comment (lbernail) [reply pkg/cloudprovider/providers/aws/aws.go:202]: Makes sense, thanks

File: pkg/cloudprovider/providers/aws/aws.go [hunk 5]
```diff
@@ -1436,51 +1446,92 @@ func (c *Cloud) InstanceType(ctx context.Context, nodeName types.NodeName) (stri
...
 // getCandidateZonesForDynamicVolume retrieves  a list of all the zones in which nodes are running
-// It currently involves querying all instances
 func (c *Cloud) getCandidateZonesForDynamicVolume() (sets.String, error) {
 	// We don't currently cache this; it is currently used only in volume
 	// creation which is expected to be a comparatively rare occurrence.
 
 	// TODO: Caching / expose v1.Nodes to the cloud provider?
 	// TODO: We could also query for subnets, I think
+	zones, err := c.getCachedCandidateZonesForDynamicVolume()
+	if err != nil {
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:1456]: getCandidateZonesForDynamicVolume looks bit weird in terms of structure because the logic is split in two:
- a one-line call to getCachedCandidateZonesForDynamicVolume
- if it fails a similar logic in the function itself

Wouldn't it be better to also use a function for the error case?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 6]
```diff
@@ -1436,51 +1446,92 @@ func (c *Cloud) InstanceType(ctx context.Context, nodeName types.NodeName) (stri
...
+		glog.V(2).Infof("getCachedCandidateZonesForDynamicVolume() failed, falling back to legacy method: %v", err)
 
-	filters := []*ec2.Filter{newEc2Filter("instance-state-name", "running")}
+		filters := []*ec2.Filter{newEc2Filter("instance-state-name", "running")}
 
-	instances, err := c.describeInstances(filters)
+		instances, err := c.describeInstances(filters)
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:1461]: This call is really not optimal (it gets all instances in the account and filter locally)
I wonder if we shouldn't modify this function (describeInstances) instead (as the TODO shows in the existing code)
It's a bit of a detail that is probably better discussed upstream

File: pkg/cloudprovider/providers/aws/aws.go [hunk 7]
```diff
@@ -1436,51 +1446,92 @@ func (c *Cloud) InstanceType(ctx context.Context, nodeName types.NodeName) (stri
...
+			return nil, err
+		}
+
+		if len(instances) == 0 {
+			return nil, fmt.Errorf("no instances returned")
+		}
+
+		zones := sets.NewString()
+
+		for _, instance := range instances {
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:1472]: The filtering logic could be shared for both methods (but this would required generating a new object for each instance/node based on where it comes from: informer or ec2 describe API call)
It would be easier to understand I think

File: pkg/cloudprovider/providers/aws/aws.go (hunk 6 above)
Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:1461]: This is the legacy path, we don't use it if the Node-Informer solution works. Maybe this path could be completely removed if the assumption is the Node-Informer should always work. That's the discussion I hope to see upstream, indeed.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 7 above)
Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:1472]: Again, this is the legacy path, we only take it if the Node-Informer is not ready. I would favour removing the whole block if we can. Discussion for upstream.

I leave the legacy path untouched as a safe fallback. We can be more adventurous upstream.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 5 above)
Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:1456]: Done!

File: pkg/cloudprovider/providers/aws/aws.go [hunk 8]
```diff
@@ -1435,15 +1445,29 @@ func (c *Cloud) InstanceType(ctx context.Context, nodeName types.NodeName) (stri
...
 func (c *Cloud) getCandidateZonesForDynamicVolume() (sets.String, error) {
 	// We don't currently cache this; it is currently used only in volume
 	// creation which is expected to be a comparatively rare occurrence.
 
 	// TODO: Caching / expose v1.Nodes to the cloud provider?
 	// TODO: We could also query for subnets, I think
+	zones, err := c.getCandidateZonesForDynamicVolumeViaInformer()
+	if err != nil {
+		glog.V(2).Infof("getCachedCandidateZonesForDynamicVolume() failed, falling back to legacy method: %v", err)
+		zones, err = c.getCandidateZonesForDynamicVolumeViaInformer()
```
Comment (lbernail) [inline pkg/cloudprovider/providers/aws/aws.go:1458]: this should viaAWS right?

Comment (EricMountain) [reply pkg/cloudprovider/providers/aws/aws.go:1458]: Well spotted, thanks.
